tests/test_strat.py

```diff
@@ -101,18 +105,113 @@ def check_efficiency():
+    possible so that could be copied/pasted across projects.
+
+    Defaults for atol, rtol, and eps based on `torch.autograd.gradcheck`.
+    """
+
+    def convert_none_to_zeros(sequence, like_sequence):
+        return [torch.zeros_like(q) if p is None else p for p, q in zip(sequence, like_sequence)]
+
+    def flatten(sequence):
+        return torch.cat([p.reshape(-1) for p in sequence]) if len(sequence) > 0 else torch.tensor([])
+
+    if isinstance(inputs, torch.Tensor):
+        inputs = (inputs,)
+
+    if isinstance(modules, nn.Module):
+        modules = (modules,)
+
+    # Don't modify original objects.
+    modules = tuple(copy.deepcopy(m) for m in modules)
+    inputs = tuple(i.clone().requires_grad_() for i in inputs)
+    func_only_inputs = lambda *args: func(args, modules)  # noqa
+
+    # Grad wrt inputs.
+    if grad_inputs:
+        torch.autograd.gradcheck(func_only_inputs, inputs, eps=eps, atol=atol, rtol=rtol)
+
+    # Grad of grad wrt inputs.
+    if gradgrad_inputs:
+        torch.autograd.gradgradcheck(func_only_inputs, inputs, eps=eps, atol=atol, rtol=rtol)
+
+    # Grad wrt params.
+    if grad_params:
+        params = [p for m in modules for p in m.parameters() if p.requires_grad]
+        loss = func(inputs, modules)
+        framework_grad = flatten(convert_none_to_zeros(torch.autograd.grad(loss, params, create_graph=True), params))
+
+        numerical_grad = []
+        for param in params:
+            flat_param = param.reshape(-1)
+            for i in range(len(flat_param)):
+                flat_param[i] += eps  # In-place.
+                plus_eps = func(inputs, modules)
+                flat_param[i] -= eps
+
+                flat_param[i] -= eps
+                minus_eps = func(inputs, modules)
+                flat_param[i] += eps
+
+                numerical_grad.append((plus_eps - minus_eps).detach() / (2 * eps))
+                del plus_eps, minus_eps
+        numerical_grad = torch.stack(numerical_grad)
+        torch.testing.assert_allclose(numerical_grad, framework_grad, rtol=rtol, atol=atol)
+
+    # Grad of grad wrt params.
+    if gradgrad_params:
+        # Define this outside `func_high_order` so that random tensors are generated only once.
+        grad_outputs = [torch.rand_like(p) for m in modules for p in m.parameters() if p.requires_grad]
+
+        def func_high_order(inputs, modules):
+            params = [p for m in modules for p in m.parameters() if p.requires_grad]
+            grads = torch.autograd.grad(func(inputs, modules), params, create_graph=True, allow_unused=True)
+            grads = tuple(grad for grad in grads if grad is not None)
+            return sum([(grad * grad_output).sum() for grad, grad_output in zip(grads, grad_outputs)])
+
+        swiss_knife_gradcheck(func_high_order, inputs, modules, rtol=rtol, atol=atol, eps=eps, grad_params=True)
 
 
 test_dg_ga_jvp()
 check_efficiency()
-test_adjoint()
+
+test_adjoint_inputs()
+test_adjoint_params()
```

Comment:
Would it be possible to remove these calls at the bottom and just make this operate through pytest? The tests all already start with `test_` so they should be picked up. (Not sure how you want to handle `check_efficiency`, that's really more a diagnostic than a test.)

Reply:
Done.